hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/MergeIntoHoodieTableCommand.scala

```diff
@@ -301,15 +335,20 @@ case class MergeIntoHoodieTableCommand(mergeInto: MergeIntoTable) extends Hoodie
   def sourceDataset: DataFrame = {
     val resolver = sparkSession.sessionState.analyzer.resolver
 
-    val sourceTablePlan = mergeInto.sourceTable
+    val tablemetacols = mergeInto.targetTable.output.filter(a => isMetaField(a.name))
+    val joinData = sparkAdapter.createMITJoin(mergeInto.sourceTable, mergeInto.targetTable, LeftOuter, Some(mergeInto.mergeCondition), "NONE")
```

Comment:
can you help us understand where exactly this join was happening before this patch. Was it lazy before and we are making it eager now or are we doing something additional ?

Reply:
No spark join was happening before this patch. The Hudi upsert operation during tagging matches the incoming records to their existing records if they exist. That is functionally equivalent to joining the incoming records with the existing records on the recordkey. There was a restriction that we could only match on the recordkey, so we didn't need to worry about the other columns. Now that we want to join on other columns for pkless, we need to use the spark join operation which will work for any column.

Reply:
sg. can we add docs on this as well.